README.md
# Metronome

home
: https://code.martini.nu/fossil/symphony-metronome

docs
: https://martini.nu/docs/symphony-metronome

github_mirror
: https://github.com/mahlonsmith/symphony-metronome



## Description

Metronome is an interval scheduler and task runner.  It can be used locally as a
cron replacement, or as a network-wide job executor.  It is intended to be run
alongside Symphony, a Ruby AMQP event consumer.

[![Gem Version](https://badge.fury.io/rb/symphony-metronome.svg)](https://badge.fury.io/rb/symphony-metronome)

Events are stored via simple database rows, and optionally themselves managed
via AMQP events.  Interval/time values are expressed with intuitive English
phrases, ie.: 'at 2pm', or 'Starting in 20 minutes, run every 10 seconds and
then finish in 2 days', or 'execute 12 times during the next minute'.

## Synopsis

Here's an example of a simple cron clone:

```
#!ruby

require 'symphony/metronome'

Symphony.load_config

Symphony::Metronome.run do |opts, id|
	Thread.new do
		pid = fork { exec opts.delete('command') }
		Process.waitpid( pid )
	end
end
```


And here's a simplistic timed AMQP message broadcaster, using existing Symphony
connection information:

```
#!ruby

require 'symphony/metronome'

Symphony.load_config

Symphony::Metronome.run do |opts, id|
	key = opts.delete( 'routing_key' ) or next
	exchange = Symphony::Queue.amqp_exchange
	exchange.publish( 'hi from Metronome!', routing_key: key )
end
```

## Adding Actions

There are two primary components to Metronome -- getting actions into
its database, and performing some task with those actions when the time
is appropriate.

By default, Metronome will start up an AMQP listener, attached to your
Symphony exchange, and wait for new scheduling messages.  There are two
events it will take action on:

metronome.create:

	Create a new scheduled event.  The payload should be a hash.  An
	'expression' key is required, that provides the interval description.
	Anything additional is serialized to 'options', that are passed to the
	block when the interval fires.  You can populate it with anything
	your task requires to execute.

metronome.delete:

	The payload is the row ID of the action.  Metronome removes it from
	the database.

If you'd prefer not to use the AMQP listener, you can put actions into
Metronome using any database methodology you please.  When the daemon
starts up or receives a HUP signal, it will re-read and schedule out
upcoming work.


## Options

Metronome uses
[Configurability](https://rubygems.org/gems/configurability) to determine
behavior.  The configuration is a [YAML](http://www.yaml.org/) file.  It
shares AMQP configuration with Symphony, and adds metronome specific
controls in the 'metronome' key.

	metronome:
		splay: 0
		listen: true
		db: sqlite:///tmp/metronome.db


### splay

Randomize all start times for actions by this many seconds on either
side of the original execution time.  Defaults to none.

### listen

Start up an AMQP listener using Symphony configuration, for remote
administration of schedule events.  Defaults to true.

### db

A [Sequel](https://rubygems.org/gems/sequel) connection URI.  Currently,
Metronome is tested under SQLite and PostgreSQL.  Defaults to a SQLite
file at /tmp/metronome.db.


## Scheduling Examples

Note that Metronome is designed as an interval scheduler, not a
calendaring app.  It doesn't have any concepts around phrases like "next
tuesday", or "the 3rd sunday after christmas".  If that's what you're
after, check out the [chronic](http://rubygems.org/gems/chronic)
library instead.

Here are a small set of example expressions.  Feel free to use the
*metronome-exp* utility to get a feel for what Metronome anticipates.

```
in 30.5 minutes
once an hour
every 15 minutes for 2 days
at 2014-05-01
at 2014-04-01 14:00:25
at 2pm
starting at 2pm once a day
start in 1 hour from now run every 5 seconds end at 11:15pm
every other hour
run every 7th minute for a day
once a day ending in 1 week
run once a minute for an hour starting in 6 days
10 times a minute for 2 days
run 45 times every hour
30 times per day
start at 2010-01-02 run 12 times and end on 2010-01-03
starting in an hour from now run 6 times a minute for 2 hours
beginning a day from now, run 30 times per minute and finish in 2 weeks
execute 12 times during the next 2 minutes
once a minute beginning in 5 minutes
```

In general, you can use reasonably intuitive phrasings.  Capitalization,
whitespace, and punctuation doesn't matter.  When describing numbers,
use digit/integer form instead of words, ie: '1 hour' instead of 'one
hour'.


## Installation

```
gem install symphony-metronome
```

## Contributing

You can check out the source via Fossil from the following uri:

% fossil clone https://code.martini.nu/fossil/symphony-metronome

or via its GitHub mirror at:

% git clone https://github.com/mahlonsmith/Symphony-Metronome

After checking out the source, run:

    $ gem install -Ng
    $ rake setup

This will install dependencies, and do any other necessary setup for
development.

Please report any issues
[here](https://code.martini.nu/fossil/symphony-metronome/tktnew).


## Authors

- Mahlon E. Smith <mahlon@martini.nu>


## License

Copyright (c) 2014-2023 Mahlon E. Smith
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the author/s, nor the names of the project's
  contributors may be used to endorse or promote products derived from this
  software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



History.md
# Release History for Symphony-Metronome

---

## v0.3.0 [2023-03-20] Mahlon E. Smith <mahlon@martini.nu>

- Updates for Ruby 3.
- Dependency updates.
- Conversion to more modern ruby development tooling.


## v0.2.1 [2015-07-08] Mahlon E. Smith <mahlon@martini.nu>

Enhancements:

- Add "last run" timestamps for recurring events.

Fixes:

- Repair exact start times for recurring events, ie
  'starting at 2015-01-01 09:00:00 run every other minute for 2 days'


## v0.1.0 [2014-04-22] Mahlon E. Smith <mahlon@martini.nu>

Initial release.



spec/helpers.rb
#!/usr/bin/ruby
# coding: utf-8
# vim: set nosta noet ts=4 sw=4:

require 'pathname'

BASEDIR = Pathname( __FILE__ ).dirname.parent
LIBDIR  = BASEDIR + 'lib'

$LOAD_PATH.unshift( LIBDIR.to_s )

# SimpleCov test coverage reporting; enable this using the :coverage rake task
require 'simplecov' if ENV['COVERAGE']

require 'timecop'
require 'loggability'
require 'loggability/spechelpers'
require 'configurability'
require 'configurability/behavior'
require 'rspec'

require 'symphony'
require 'symphony/metronome'

Loggability.format_with( :color ) if $stdout.tty?


### RSpec helper functions.
module Loggability::SpecHelpers
end


### Mock with RSpec
RSpec.configure do |config|
	config.run_all_when_everything_filtered = true
	config.filter_run :focus
	# config.order = 'random'
	config.expect_with( :rspec )
	config.mock_with( :rspec ) do |mock|
		mock.syntax = :expect
	end

	config.include( Loggability::SpecHelpers )
end



spec/symphony/metronome_spec.rb

#!/usr/bin/env rspec

require_relative '../helpers'


describe Symphony::Metronome do

	before( :each ) do
		ENV.delete( 'METRONOME_CONFIG' )
	end


	it "will load a default config file if none is specified" do
		config_object = double( "Configurability::Config object" )
		expect( Configurability ).to receive( :gather_defaults ).
			and_return( {} )
		expect( Configurability::Config ).to receive( :load ).
			with( described_class::DEFAULT_CONFIG_FILE, {} ).
			and_return( config_object )
		expect( config_object ).to receive( :install )

		described_class.load_config
	end


	it "will load a config file given in an environment variable if none is specified" do
		ENV['METRONOME_CONFIG'] = '/usr/local/etc/config.yml'

		config_object = double( "Configurability::Config object" )
		expect( Configurability ).to receive( :gather_defaults ).
			and_return( {} )
		expect( Configurability::Config ).to receive( :load ).
			with( '/usr/local/etc/config.yml', {} ).
			and_return( config_object )
		expect( config_object ).to receive( :install )

		described_class.load_config
	end


	it "will load a config file and install it if one is given" do
		config_object = double( "Configurability::Config object" )
		expect( Configurability ).to receive( :gather_defaults ).
			and_return( {} )
		expect( Configurability::Config ).to receive( :load ).
			with( 'a/configfile.yml', {} ).
			and_return( config_object )
		expect( config_object ).to receive( :install )

		described_class.load_config( 'a/configfile.yml' )
	end


	it "will override default values when loading the config if they're given" do
		config_object = double( "Configurability::Config object" )
		expect( Configurability ).to_not receive( :gather_defaults )
		expect( Configurability::Config ).to receive( :load ).
			with( 'a/different/configfile.yml', {database: {dbname: 'test'}} ).
			and_return( config_object )
		expect( config_object ).to receive( :install )

		described_class.load_config( 'a/different/configfile.yml', database: {dbname: 'test'} )
	end
end



spec/symphony/metronome/mixins_spec.rb
#!/usr/bin/env rspec -wfd
# vim: set nosta noet ts=4 sw=4:

require_relative '../../helpers'

using Symphony::Metronome::TimeRefinements


describe Symphony::Metronome, 'mixins' do

	describe "numeric constant methods" do

		SECONDS_IN_A_MINUTE    = 60
		SECONDS_IN_AN_HOUR     = SECONDS_IN_A_MINUTE * 60
		SECONDS_IN_A_DAY       = SECONDS_IN_AN_HOUR * 24
		SECONDS_IN_A_WEEK      = SECONDS_IN_A_DAY * 7
		SECONDS_IN_A_FORTNIGHT = SECONDS_IN_A_WEEK * 2
		SECONDS_IN_A_MONTH     = SECONDS_IN_A_DAY * 30
		SECONDS_IN_A_YEAR      = Integer( SECONDS_IN_A_DAY * 365.25 )

		it "can calculate the number of seconds for various units of time" do
			expect( 1.second ).to eq( 1 )
			expect( 14.seconds ).to eq( 14 )

			expect( 1.minute ).to eq( SECONDS_IN_A_MINUTE )
			expect( 18.minutes ).to eq( SECONDS_IN_A_MINUTE * 18 )

			expect( 1.hour ).to eq( SECONDS_IN_AN_HOUR )
			expect( 723.hours ).to eq( SECONDS_IN_AN_HOUR * 723 )

			expect( 1.day ).to eq( SECONDS_IN_A_DAY )
			expect( 3.days ).to eq( SECONDS_IN_A_DAY * 3 )

			expect( 1.week ).to eq( SECONDS_IN_A_WEEK )
			expect( 28.weeks ).to eq( SECONDS_IN_A_WEEK * 28 )

			expect( 1.fortnight ).to eq( SECONDS_IN_A_FORTNIGHT )
			expect( 31.fortnights ).to eq( SECONDS_IN_A_FORTNIGHT * 31 )

			expect( 1.month ).to eq( SECONDS_IN_A_MONTH )
			expect( 67.months ).to eq( SECONDS_IN_A_MONTH * 67 )

			expect( 1.year ).to eq( SECONDS_IN_A_YEAR )
			expect( 13.years ).to eq( SECONDS_IN_A_YEAR * 13 )
		end


		it "can calculate various time offsets" do
			starttime = Time.now

			expect( 1.second.after( starttime ) ).to eq( starttime + 1 )
			expect( 18.seconds.from_now ).to be_within( 10.seconds ).of( starttime + 18 )

			expect( 1.second.before( starttime ) ).to eq( starttime - 1 )
			expect( 3.hours.ago ).to be_within( 10.seconds ).of( starttime - 10800 )
		end
	end
end



lib/symphony/metronome.rb
#!/usr/bin/env ruby
# vim: set nosta noet ts=4 sw=4:

require 'pathname'
require 'symphony' unless defined?( Symphony )

module Symphony::Metronome
	extend Loggability,
	       Configurability

	# Library version constant
	VERSION = '0.3.0'

	# The name of the environment variable to check for config file overrides
	CONFIG_ENV = 'METRONOME_CONFIG'

	# The path to the default config file
	DEFAULT_CONFIG_FILE = 'etc/config.yml'

	# The data directory that contains migration files.
	#
	DATADIR = if ENV['METRONOME_DATADIR']
				   Pathname( ENV['METRONOME_DATADIR'] )
			   elsif Gem.loaded_specs[ 'symphony-metronome' ] && File.exist?( Gem.loaded_specs['symphony-metronome'].datadir )
				   Pathname( Gem.loaded_specs['symphony-metronome'].datadir )
			   else
				   Pathname( __FILE__ ).dirname.parent.parent + 'data/symphony-metronome'
			   end


	# Loggability API -- use symphony's logger
	log_to :symphony

	# Configurability API
	config_key :metronome


	### Get the loaded config (a Configurability::Config object)
	def self::config
		Configurability.loaded_config
	end


	### Load the specified +config_file+, install the config in all objects with
	### Configurability, and call any callbacks registered via #after_configure.
	def self::load_config( config_file=nil, defaults=nil )
		config_file ||= ENV[ CONFIG_ENV ] || DEFAULT_CONFIG_FILE
		defaults    ||= Configurability.gather_defaults
		config = Configurability::Config.load( config_file, defaults )
		config.install
	end


	# The generic parse exception class.
	class TimeParseError < ArgumentError; end

	require 'symphony/metronome/scheduler'
	require 'symphony/metronome/intervalexpression'
	require 'symphony/metronome/scheduledevent'
	require 'symphony/tasks/scheduletask'


	###############
	module_function
	###############

	### Convenience method for running the scheduler.
	###
	def run( &block )
		raise LocalJumpError, "No block provided." unless block_given?
		return Symphony::Metronome::Scheduler.run( &block )
	end

end # Symphony::Metronome



lib/symphony/metronome/mixins.rb
# -*- ruby -*-
#encoding: utf-8
# vim: set nosta noet ts=4 sw=4:

require 'symphony' unless defined?( Symphony )
require 'symphony/metronome' unless defined?( Symphony::Metronome )


module Symphony::Metronome

	# Functions for time calculations
	module TimeFunctions

		###############
		module_function
		###############

		### Calculate the (approximate) number of seconds that are in +count+ of the
		### given +unit+ of time.
		###
		def calculate_seconds( count, unit )
			return case unit
			when :seconds, :second
				count
			when :minutes, :minute
				count * 60
			when :hours, :hour
				count * 3600
			when :days, :day
				count * 86400
			when :weeks, :week
				count * 604800
			when :fortnights, :fortnight
				count * 1209600
			when :months, :month
				count * 2592000
			when :years, :year
				count * 31557600
			else
				raise ArgumentError, "don't know how to calculate seconds in a %p" % [ unit ]
			end
		end
	end # module TimeFunctions


	# Refinements to Numeric to add time-related convenience methods
	module TimeRefinements
		refine Numeric do

			### Number of seconds (returns receiver unmodified)
			def seconds
				return self
			end
			alias_method :second, :seconds

			### Returns number of seconds in <receiver> minutes
			def minutes
				return TimeFunctions.calculate_seconds( self, :minutes )
			end
			alias_method :minute, :minutes

			### Returns the number of seconds in <receiver> hours
			def hours
				return TimeFunctions.calculate_seconds( self, :hours )
			end
			alias_method :hour, :hours

			### Returns the number of seconds in <receiver> days
			def days
				return TimeFunctions.calculate_seconds( self, :day )
			end
			alias_method :day, :days

			### Return the number of seconds in <receiver> weeks
			def weeks
				return TimeFunctions.calculate_seconds( self, :weeks )
			end
			alias_method :week, :weeks

			### Returns the number of seconds in <receiver> fortnights
			def fortnights
				return TimeFunctions.calculate_seconds( self, :fortnights )
			end
			alias_method :fortnight, :fortnights

			### Returns the number of seconds in <receiver> months (approximate)
			def months
				return TimeFunctions.calculate_seconds( self, :months )
			end
			alias_method :month, :months

			### Returns the number of seconds in <receiver> years (approximate)
			def years
				return TimeFunctions.calculate_seconds( self, :years )
			end
			alias_method :year, :years


			### Returns the Time <receiver> number of seconds before the
			### specified +time+. E.g., 2.hours.before( header.expiration )
			def before( time )
				return time - self
			end


			### Returns the Time <receiver> number of seconds ago. (e.g.,
			### expiration > 2.hours.ago )
			def ago
				return self.before( ::Time.now )
			end


			### Returns the Time <receiver> number of seconds after the given +time+.
			### E.g., 10.minutes.after( header.expiration )
			def after( time )
				return time + self
			end


			### Return a new Time <receiver> number of seconds from now.
			def from_now
				return self.after( ::Time.now )
			end

		end # refine Numeric
	end # module TimeRefinements

end # module Symphony::Metronome




lib/symphony/metronome/scheduler.rb
#!/usr/bin/env ruby
# vim: set nosta noet ts=4 sw=4:

require 'symphony'
require 'symphony/metronome'


### Manage the delta queue of events and associated actions.
###
class Symphony::Metronome::Scheduler
	extend Loggability, Configurability
	include Symphony::SignalHandling

	log_to :symphony
	config_key :metronome

	# Signals the daemon responds to.
	SIGNALS = [ :HUP, :INT, :TERM ]


	### Configurability API.
	###
	configurability do

		# Should Metronome register and schedule events via AMQP?
		# If +false+, you'll need a separate way to add event actions
		# to the database, and manually HUP the daemon.
		setting :listen, default: false
	end


	### Create and start an instanced daemon.
	###
	def self::run( &block )
		return new( block )
	end


	### Actions to perform when creating a new daemon.
	###
	private_class_method :new
	def initialize( block ) #:nodoc:

		# Start the queue subscriber for schedule changes.
		#
		if self.class.listen
			Symphony::Metronome::ScheduledEvent.db.disconnect
			@child = fork do
				$0 = 'Metronome (listener)'
				Symphony::Metronome::ScheduleTask.run
			end
			Process.setpgid( @child, 0 )
		end

		# Signal handling for the master (this) process.
		#
		self.set_up_signal_handling
		self.set_signal_traps( *SIGNALS )

		@queue = Symphony::Metronome::ScheduledEvent.load
		@proc  = block

		# Enter the main loop.
		self.start

	rescue => err
		self.log.error "%p while running: %s" % [ err.class, err.message ]
		self.log.debug "  " + err.backtrace.join( "\n  " )
		Process.kill( 'TERM', @child ) if self.class.listen
	end


	# The sorted set of ScheduledEvent objects.
	attr_reader :queue


	#########
	protected
	#########

	### Main daemon sleep loop.
	###
	def start
		$0 = "Metronome%s" % [ self.class.listen ? ' (executor)' : '' ]
		@running = true

		loop do
			wait = nil

			if ev = self.queue.first
				wait = ev.runtime - Time.now
				wait = 0 if wait < 0
				self.log.info "Next event in %0.3f second(s) (id: %d)..." % [ wait, ev.id ]
			else
				self.log.warn "No events scheduled.  Waiting indefinitely..."
			end

			self.process_events unless self.wait_for_signals( wait )
			break unless @running
		end
	end


	### Dispatch incoming signals to appropriate handlers.
	###
	def handle_signal( sig )
		case sig
		when :TERM, :INT
			@running = false
			Process.kill( sig.to_s, @child ) if self.class.listen

		when :HUP
			@queue = Symphony::Metronome::ScheduledEvent.load
			self.queue.each{|ev| ev.fire(&@proc) if ev.event.recurring }

		else
			self.log.debug "Unhandled signal: %s" % [ sig ]
		end
	end


	### Process all events that have reached their runtime.
	###
	def process_events
		now = Time.now

		events_to_delete = []
		events_to_add = []
		self.queue.each do |ev|
			next unless now >= ev.runtime

			events_to_delete << ev
			rv = ev.fire( &@proc )

			# Reschedule the event and place it back on the queue.
			#
			if ev.event.recurring
				ev.reset_runtime
				events_to_add << ev unless rv.nil?

			# It was a single run event, torch it!
			#
			else
				ev.delete

			end
		end

		events_to_delete.map{|e| self.queue.delete( e ) }
		events_to_add.map{|e| self.queue.add( e ) }
	end

end # Symphony::Metronome::Scheduler



lib/symphony/metronome/scheduledevent.rb
#!/usr/bin/env ruby
# vim: set nosta noet ts=4 sw=4:

require 'set'
require 'sequel'
require 'sqlite3'
require 'yajl'
require 'symphony/metronome'

Sequel.extension :migration


### A class the represents the relationship between an interval and
### an event.
###
class Symphony::Metronome::ScheduledEvent
	extend Loggability, Configurability
	include Comparable

	log_to :symphony
	config_key :metronome


	### Configurability API.
	###
	configurability do

		##
		# A Sequel-style DB connection URI.
		setting :db, default: 'sqlite:///tmp/metronome.db'

		##
		# Adjust recurring intervals by a random window.
		setting :splay, default: 0
	end


	######################################################################
	# C L A S S  M E T H O D S
	######################################################################

	### Configurability API.
	###
	def self::configure( config=nil )
		config = self.defaults.merge( config || {} )
		@db    = Sequel.connect( config.delete(:db) )
		@splay = config.delete( :splay )

		# Ensure the database is current.
		#
		migrations_dir = Symphony::Metronome::DATADIR + 'migrations'
		unless Sequel::Migrator.is_current?( self.db, migrations_dir.to_s )
			Loggability[ Symphony ].info "Installing database schema..."
			Sequel::Migrator.apply( self.db, migrations_dir.to_s )
		end
	end


	### Return a set of all known events, sorted by date of execution.
	### Delete any rows that are invalid expressions.
	###
	def self::load
		events = SortedSet.new

		# Force reset the DB handle.
		self.db.disconnect

		self.log.debug "Parsing/loading all actions."
		self.db[ :metronome ].each do |event|
			begin
				event = new( event )
				events << event
			rescue ArgumentError, Symphony::Metronome::TimeParseError => err
				self.log.error "%p while parsing \"%s\": %s" % [
					err.class,
					event[:expression],
					err.message
				]
				self.log.debug "  " + err.backtrace.join( "\n  " )
				self.db[ :metronome ].filter( :id => event[:id] ).delete
			end
		end

		return events
	end


	######################################################################
	# I N S T A N C E  M E T H O D S
	######################################################################

	### Create a new ScheduledEvent object.
	###
	def initialize( row )
		@event    = Symphony::Metronome::IntervalExpression.parse( row[:expression], row[:created] )
		@options  = row.delete( :options )
		@id       = row.delete( :id )
		@ds       = self.class.db[ :metronome ].filter( :id => self.id )

		self.reset_runtime

		unless self.class.splay.zero?
			splay = Range.new( - self.class.splay, self.class.splay )
			@runtime = self.runtime + rand( splay )
		end
	end

	# The sequel dataset representing this event.
	attr_reader :ds

	# The parsed interval expression.
	attr_reader :event

	# The unique ID number of the scheduled event.
	attr_reader :id

	# The options hash attached to this event.
	attr_reader :options

	# The exact time that this event will run.
	attr_reader :runtime


	### Set the datetime that this event should fire next.
	###
	def reset_runtime
		now = Time.now

		# Start time is in the future, so it's sufficent to be considered the run time.
		#
		if self.event.starting >= now
			@runtime = self.event.starting
			return
		end

		# Otherwise, the event should already be running (start time has already
		# elapsed), so schedule it forward on it's next interval iteration.
		#
		# If it's a recurring event that has run before, consider the elapsed time
		# as part of the next calculation.
		#
		row = self.ds.first
		if self.event.recurring && row
			last = row[ :lastrun ]
			if last && now > last
				@runtime = now + self.event.interval - ( now - last )
			else
				@runtime = now + self.event.interval
			end

		else
			@runtime = now + self.event.interval
		end
	end


	### Perform the action attached to the event.  Yields the
	### deserialized options, the action ID to the supplied block if
	### this event is okay to execute.
	###
	### If the event is recurring, perform additional checks against the
	### last run time.
	###
	### Automatically remove the event if it has expired.
	###
	def fire
		rv = self.event.fire?

		# Just based on the expression parser, is this event ready to fire?
		#
		if rv
			opts = Yajl.load( self.options )

			# Don't fire recurring events unless their interval has elapsed.
			# This prevents events from triggering when the daemon receives
			# a HUP.
			#
			if self.event.recurring
				now = Time.now
				row = self.ds.first

				if row
					last = row[ :lastrun ]
					return false if last && now - last < self.event.interval
				end

				# Mark the time this recurring event was fired.
				self.ds.update( :lastrun => Time.now )
			end

			yield opts, self.id
		end

		self.delete if rv.nil?
		return rv
	end


	### Permanently remove this event from the database.
	###
	def delete
		self.log.debug "Removing action %p" % [ self.id ]
		self.ds.delete
	end


	### Comparable interface, order by next run time, soonest first.
	###
	def <=>( other )
		return self.runtime <=> other.runtime
	end

end # Symphony::Metronome::ScheduledEvent



lib/symphony/tasks/scheduletask.rb
#!/usr/bin/env ruby
# vim: set nosta noet ts=4 sw=4:

require 'symphony'
require 'symphony/routing'
require 'symphony/metronome'


### Accept metronome scheduling events, translating them
### to DB rows for persistence.
###
class Symphony::Metronome::ScheduleTask < Symphony::Task
	include Symphony::Routing

	queue_name 'metronome'
	timeout 30

	### Get a handle to the database.
	###
	def initialize( * )
		@db = Symphony::Metronome::ScheduledEvent.db
		@actions = @db[ :metronome ]
		super
	end

	# The Sequel dataset of scheduled event actions.
	attr_reader :actions


	### Accept a new scheduled event.  The payload should be a free
	### form hash of options, along with an expression string that
	### conforms to IntervalExpression.
	###
	###   {
	###       :expression  => 'run 25 times for an hour',
	###       :payload     => { ... },
	###   }
	###
	on 'metronome.create' do |payload, metadata|
		raise ArgumentError, 'Invalid payload.' unless payload.is_a?( Hash )
		exp = payload.delete( 'expression' )
		raise ArgumentError, 'Missing time expression.' unless exp

		self.actions.insert(
			:created     => Time.now,
			:expression  => exp,
			:options     => Yajl.dump( payload )
		)

		self.signal_parent
		return true
	end


	### Delete an existing scheduled event.
	### The payload is the id of the action (row) to delete.
	###
	on 'metronome.delete' do |id, metadata|
		self.actions.filter( :id => id.to_i ).delete
		self.signal_parent
		return true
	end


	### Tell our parent (the Metronome broadcaster) to re-read its event
	### list.
	###
	def signal_parent
		parent = Process.ppid

		# Check to make sure we weren't orphaned.
		#
		if parent == 1
			self.log.error "Lost my parent process?  Exiting."
			exit 1
		end

		Process.kill( 'HUP', parent )
	end
end

